src/librbd/migration/NBDStream.cc

```diff
@@ -0,0 +1,278 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/migration/NBDStream.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "librbd/AsioEngine.h"
+#include "librbd/ImageCtx.h"
+#include <boost/asio/buffer.hpp>
+
+#undef FMT_HEADER_ONLY
+#define FMT_HEADER_ONLY 1
+#include <fmt/chrono.h>
+#include <fmt/format.h>
+
+#include <time.h>
+
+namespace librbd {
+namespace migration {
+
+namespace {
+
+const std::string SERVER_KEY {"server"};
+const std::string PORT_KEY {"port"};
+
+} // anonymous namespace
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::migration::NBDStream: " << this \
+                           << " " << __func__ << ": "
+
+int check_extent(void *data, 
+                 const char *metacontext,
+                 uint64_t offset,
+                 uint32_t *entries, size_t nr_entries, int *error) {
+  io::SparseExtents* sparse_extents = (io::SparseExtents*)data;
+  uint64_t length = 0;
+  for (size_t i=0; i<nr_entries; i+=2) {
+    length += entries[i];
+  }
+  auto state = io::SPARSE_EXTENT_STATE_DATA;
+  if (nr_entries == 2) {
+    if (entries[1] & (LIBNBD_STATE_HOLE | LIBNBD_STATE_ZERO)) {
+      state = io::SPARSE_EXTENT_STATE_ZEROED;
+    }
+  }
+  sparse_extents->insert(offset, length, {state, length});
+  return 1;
+}
+
+int task_completed (void *user_data, int *error)
+{
+  Context* on_finish = (Context *)user_data;
+  on_finish->complete(0);
+  return 1;
+}
+
+template <typename I>
+struct NBDStream<I>::ReadRequest {
+  NBDStream*  nbd_stream;
+  io::Extents byte_extents;
+  bufferlist* data;
+  Context* on_finish;
+
+  ReadRequest(NBDStream* nbd_stream, io::Extents&& byte_extents,
+              bufferlist* data, Context* on_finish)
+    : nbd_stream(nbd_stream), byte_extents(std::move(byte_extents)),
+      data(data), on_finish(on_finish) {
+    auto cct = nbd_stream->m_cct;
+    ldout(cct, 20) << dendl;
+  }
+
+  void send_read() {
+    data->clear();
+    read();
+  }
+
+  void read() {
+    auto cct = nbd_stream->m_cct;  
+    struct nbd_handle *nbd = nbd_stream->nbd;
+
+    ldout(cct, 20) << "byte_extents=" << byte_extents << dendl;
+
+    for (auto [byte_offset, byte_length] : byte_extents) {
+      ldout(cct, 20) << "byte_offset=" << byte_offset << dendl;
+      ldout(cct, 20) << "byte_length=" << byte_length << dendl;
+      auto ptr = buffer::ptr_node::create(buffer::create_small_page_aligned(
+        byte_length));
+      auto buffer = boost::asio::mutable_buffer(ptr->c_str(), byte_length);
+      data->push_back(std::move(ptr));
+      auto rc = nbd_aio_pread(nbd, boost::asio::buffer_cast<void *>(buffer),
+        byte_length, byte_offset, 
+        (nbd_completion_callback) { .callback=task_completed,
+                                    .user_data=on_finish }, 0);
```

Comment:
I am still in the process of trying it but here is one thing that I just noticed is definitely wrong. When I suggested to just provide `on_finish` as `user_data` and complete it in the callback, I missed that we are doing it in the `for` loop for every extent. And we may compete it only once when all callbacks are completed. So it should be a bit more complicated. E.g. ` C_Gather` may be used (you may find many examples in librbd source), or you could implement your own "completion" class.

Reply:
I see what you mean.  Good catch.  I'll see what I can do to handle it better.  BTW, I hope you and you loved ones are safe.

Reply:
Hi @trociny I hope all is well. I recognize that you might have more pressing issues to deal with, so take your time. I pushed back the original handling of the on_finish which calls it once the requests are fully served.  The current version uses a boost asio strand to avoid blocking the main caller.  I am not certain we really gain much benefit from the using the async version of the api other than possibly batching them.  And if I look at the file stream handling, I see that there a strand is also used, but the data is read serially.  We can switch back to that model if you prefer it.